services/src/main/java/org/keycloak/social/apple/AppleIdentityProvider.java

```diff
@@ -0,0 +1,264 @@
+                                     @FormParam(AppleIdentityProvider.OAUTH2_PARAMETER_USER) String user,
+                                     @FormParam(OAuth2Constants.ERROR) String error) {
+            if (error != null) {
+                logger.error(error + " for broker login " + getConfig().getProviderId());
+                if (error.equals(ACCESS_DENIED)) {
+                    return callback.cancelled(state);
+                } else if (error.equals(OAuthErrorException.LOGIN_REQUIRED) || error.equals(OAuthErrorException.INTERACTION_REQUIRED)) {
+                    return callback.error(state, error);
+                } else {
+                    return callback.error(state, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
+                }
+            }
+
+            try {
+                if (authorizationCode != null) {
+
+                    String response = generateTokenRequest(authorizationCode).asString();
+
+                    BrokeredIdentityContext federatedIdentity = getFederatedIdentity(user, response);
+                    federatedIdentity.setIdpConfig(getConfig());
+                    federatedIdentity.setIdp(AppleIdentityProvider.this);
+                    federatedIdentity.setCode(state);
+                    return callback.authenticated(federatedIdentity);
+                }
+            } catch (WebApplicationException e) {
+                return e.getResponse();
+            } catch (Exception e) {
+                logger.error("Failed to make identity provider oauth callback", e);
+            }
+            event.event(EventType.LOGIN);
+            event.error(Errors.IDENTITY_PROVIDER_LOGIN_FAILURE);
+            return ErrorPage.error(session, null, Response.Status.BAD_GATEWAY, Messages.IDENTITY_PROVIDER_UNEXPECTED_ERROR);
+        }
+
+        public BrokeredIdentityContext getFederatedIdentity(String userData, String response) throws JsonProcessingException {
+            JsonNode encodedIdToken = mapper.readTree(response).get("id_token");
+
+            JsonWebToken idToken = validateToken(encodedIdToken.asText(), false);
+
+            BrokeredIdentityContext user = new BrokeredIdentityContext(idToken.getSubject());
+            user.setUsername(idToken.getSubject());
+            if (userData !=null) {
+                JsonNode profile = mapper.readTree(userData);
+                String email = profile.get("email").asText();
+
+                String firstName = profile.get("name").get("firstName").asText();
+                String lastName = profile.get("name").get("lastName").asText();
+
+                user.setEmail(email);
+                user.setFirstName(firstName);
+                user.setLastName(lastName);
+                user.setUsername(firstName + " " + lastName);
+                AbstractJsonUserAttributeMapper.storeUserProfileForMapper(user, profile, getConfig().getAlias());
+            }
+
+            return user;
```

Comment:
The returned BrokeredIdentityContext doesn't contain the nonce in the context data, which causes the flow to fail at a later stage inside OIDCIdentityProvider.preprocessFederatedIdentity() where a nonce is expected to be present.

Also, I am wondering why you are not simply reusing the OIDCIdentityProvider.getFederatedIdentity() superclass method, as it deals with both issues i am mentioning here.

Reply:
Maybe you could reuse OIDCIdentityProvider.getFederatedIdentity() like so?
```
        public BrokeredIdentityContext getFederatedIdentity(String userData, String response) throws JsonProcessingException {
            BrokeredIdentityContext user = AppleIdentityProvider.this.getFederatedIdentity(response);

            if (userData != null) {
                JsonNode profile = mapper.readTree(userData);

                JsonNode nameNode = profile.get("name");
                if (nameNode != null) {
                    JsonNode firstNameNode = nameNode.get("firstName");
                    if (firstNameNode != null) {
                        user.setFirstName(firstNameNode.asText());
                    }
                    JsonNode lastNameNode = nameNode.get("lastName");
                    if (lastNameNode != null) {
                        user.setLastName(lastNameNode.asText());
                    }
                }

                AbstractJsonUserAttributeMapper.storeUserProfileForMapper(user, profile, getConfig().getAlias());
            }

            return user;
        }
```